src/Spice86.Core/Emulator/Devices/Sound/SoftwareMixer.cs

```diff
@@ -24,54 +23,59 @@ public sealed class SoftwareMixer : IDisposable, IDebuggableComponent {
+        Channels = _channels.ToFrozenDictionary();
     }
-
-
+    
     /// <summary>
     /// Gets the sound channels in a read-only dictionary.
     /// </summary>
-    public IDictionary<SoundChannel, AudioPlayer> Channels => _channelsReadOnlyDictionary;
+    public FrozenDictionary<SoundChannel, AudioPlayer> Channels { get; private set; } = new Dictionary<SoundChannel, AudioPlayer>().ToFrozenDictionary();
 
     internal int Render(Span<float> data, SoundChannel channel) {
         if (channel.Volume == 0 || channel.IsMuted) {
             _channels[channel].WriteSilence();
             return data.Length;
         }
-        ApplyStereoSeparation(data, channel);
-        ApplyVolume(data, channel);
-        return _channels[channel].WriteData(data);
-    }
-
-    internal int Render(Span<int> data, SoundChannel channel) {
-        if (channel.Volume == 0 || channel.IsMuted) {
-            _channels[channel].WriteSilence();
-            return data.Length;
+        float volumeFactor = channel.Volume / 100f;
+        float separation = channel.StereoSeparation / 100f;
+        
+        Span<float> target = stackalloc float[data.Length];
+        for (int i = 0; i < data.Length; i++) {
+            target[i] = data[i]  * volumeFactor * (1 - separation);
         }
-        ApplyStereoSeparation(data, channel);
-        ApplyVolume(data, channel);
-        return _channels[channel].WriteData(data);
+        return _channels[channel].WriteData(target);
     }
-
+    
     internal int Render(Span<short> data, SoundChannel channel) {
         if (channel.Volume == 0 || channel.IsMuted) {
             _channels[channel].WriteSilence();
             return data.Length;
         }
-        ApplyStereoSeparation(data, channel);
-        ApplyVolume(data, channel);
-        return _channels[channel].WriteData(data);
+        float volumeFactor = channel.Volume / 100f;
+        float separation = channel.StereoSeparation / 100f;
+
+        Span<float> target = stackalloc float[data.Length];
+        for (int i = 0; i < data.Length; i++) {
+            target[i] = (data[i] / 32768f)  * volumeFactor * (1 - separation);
+        }
+
+        return _channels[channel].WriteData(target);
     }
     
     internal int Render(Span<byte> data, SoundChannel channel) {
         if (channel.Volume == 0 || channel.IsMuted) {
             _channels[channel].WriteSilence();
             return data.Length;
         }
-        ApplyStereoSeparation(data, channel);
-        ApplyVolume(data, channel);
-        return _channels[channel].WriteData(data);
+        float volumeFactor = channel.Volume / 100f;
+        float separation = channel.StereoSeparation / 100f;
+        
+        Span<float> target = stackalloc float[data.Length];
+        for (int i = 0; i < data.Length; i++) {
+            target[i] = ((data[i] - 127) / 128f)  * volumeFactor * (1 - separation);
```

Comment:
Same thing here, move the calculation of the multiplier outside the loop

Reply:
Done.